packages/react-server/src/ReactFizzServer.js

```diff
@@ -97,119 +212,580 @@ function renderNode(request: Request, node: ReactNodeList): void {
   if (typeof type === 'function') {
     try {
       const result = type(props);
-      renderNode(result);
+      renderNode(request, parentBoundary, segment, result);
     } catch (x) {
       if (typeof x === 'object' && x !== null && typeof x.then === 'function') {
         // Something suspended, we'll need to create a new segment and resolve it later.
-        request.pendingChunks++;
-        const newSegment = createSegment(request, node);
-        const ping = newSegment.ping;
+        const insertionIndex = segment.chunks.length;
+        const newSegment = createPendingSegment(request, insertionIndex, null);
+        const suspendedWork = createSuspendedWork(
+          request,
+          node,
+          parentBoundary,
+          newSegment,
+          null,
+        );
+        const ping = suspendedWork.ping;
         x.then(ping, ping);
         // TODO: Emit place holder
       } else {
-        reportError(request, x);
-        // TODO: Emit client render signal
+        // We can rethrow to terminate the rest of this tree.
+        throw x;
       }
     }
   } else if (typeof type === 'string') {
-    request.completedChunks.push(formatChunk(type, props));
+    // TODO: Split this to start/end and process children.
+    segment.chunks.push(formatChunk(type, props));
+  } else if (type === REACT_SUSPENSE_TYPE) {
+    // Each time we enter a suspense boundary, we split out into a new segment for
+    // the fallback so that we can later replace that segment with the content.
+    // This also lets us split out the main content even if it doesn't suspend,
+    // in case it ends up generating a large subtree of content.
+    const fallback: ReactNodeList = props.fallback;
+    const content: ReactNodeList = props.children;
+
+    const newBoundary = createSuspenseBoundary();
+
+    const insertionIndex = segment.chunks.length;
```

Comment:
Do I understand correctly that you're never using indexes for DOM manipulation, and it's just something used while building a tree? Thinking of a case where some browser extension 
messes with the DOM before we've had a chance to load all the follow-up content. I guess that would break hydration anyway..

Reply:
Yea, this index is completely internal to the Fizz data structures. It's just a way to model that the child Segments are in the middle of the list of Chunks. It could also be modeled as `Array<Segment | Chunk>` but it's slightly less efficient.